### ISSUE TITLE
NA SM: Mercury busy spinning indefinitely after servicing a single client

### ISSUE THREAD

**Issue by carns** [mercury-hpc/mercury#317]
**Describe the bug**

Mercury seems to consume 100% CPU despite a non-zero timeout value in HG_Progress() once a single client has completed an RPC and exited.

**To Reproduce**

* start a server daemon using na+sm or ofi+tcp (have not tested other transports)
* have library or code atop mercury (like margo) use an HG_Progress() timeout value of 100
* observe that the server consumes minimal CPU
* run a test client that issues and completes a trivial RPC and exits
* observe that the server now consumes 100% CPU

The margo library briefly toggles to a timeout value of 0 while servicing the RPC as part of a simple adaptive strategy, so a CPU spike is expected, but it should return back to an idle state once the RPC is complete and the timout is again set to 100.

**Expected behavior**

Expected CPU usage to be minimal once server becomes quiescent again.

**Platform (please complete the following information):**
 - Ubuntu 19.10
 - gcc 8.3.0
 - libfabric 1.8.1 (when libfabric is used)

**Additional context**

**Comment by carns** [comment]
I'll try to come up with either a more detailed analysis or else a Mercury-only reproducer, just wanted to go ahead and share issue in case anyone has ideas.

**Comment by soumagne** [comment]
Hmm ok we also do something similar inside mercury since we use the try_wait() NA callback to make sure we don't sleep if there's still work to do. It should also report that it is safe to sleep in that case though...

**Comment by carns** [comment]
This can be reproduced with the existing test programs if you make a minor modification.  To reproduce:

1. comment out https://github.com/mercury-hpc/mercury/blob/master/Testing/mercury_test.c#L247 (to make it so that the test programs do not automatically shut down the test server)
2. build mercury with tests
3. run `./hg_test_server -c na -p sm -L`
4. observe cpu usage in top and confirm that hg_test_server is not at the top of the list
```
[pcarns@carns-x1 bin]$ top -b -n 1 |head -n 10
top - 21:29:35 up  7:43,  1 user,  load average: 0.75, 0.73, 0.38
Tasks: 286 total,   1 running, 284 sleeping,   1 stopped,   0 zombie
%Cpu(s):  3.0 us,  1.5 sy,  0.0 ni, 95.5 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem :   7852.3 total,   3084.9 free,   2209.5 used,   2557.9 buff/cache
MiB Swap:    980.0 total,    534.0 free,    446.0 used.   4716.3 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
 5136 pcarns    20   0   12112   3912   3228 R  13.3   0.0   0:00.02 top
   19 root      20   0       0      0      0 I   6.7   0.0   0:11.32 kworker/1+
 1122 root      20   0  238172   3904   3192 S   6.7   0.0   0:00.82 accounts-+
```
5. run `./hg_test_rpc -c na -p sm`
6. check top again
```
[pcarns@carns-x1 bin]$ top -b -n 1 |head -n 10
top - 21:31:03 up  7:44,  1 user,  load average: 0.54, 0.69, 0.40
Tasks: 286 total,   1 running, 284 sleeping,   1 stopped,   0 zombie
%Cpu(s): 15.2 us, 12.1 sy,  0.0 ni, 69.7 id,  0.0 wa,  0.0 hi,  3.0 si,  0.0 st
MiB Mem :   7852.3 total,   3055.5 free,   2219.8 used,   2577.0 buff/cache
MiB Swap:    980.0 total,    534.0 free,    446.0 used.   4687.1 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
 5125 pcarns    20   0  687580  22776   2244 S 106.7   0.3   0:04.25 hg_test_s+
 1207 root      20   0       0      0      0 I   6.7   0.0   0:00.89 kworker/2+
 2411 pcarns    20   0 3141624 208036  71444 S   6.7   2.6  13:52.44 gnome-she+
```

The hg_test_server executable stays at or near 100% CPU usage from then on.

**Comment by carns** [comment]
After adding some debugging printfs I can see that after completion of the client rpcs, the server is continuously calling epoll_wait() (with a timeout, though the function returns immediately) in na_sm, not in the mercury_core code.

Is it possible that the file descriptor associated with the client that previously connected is not being removed from poll set, or else is being left with a flag for POLLOUT that will never be cleared?

I can debug further, but @soumagne you probably have a better idea of what's supposed to be happening in here once a client exits.

**Comment by soumagne** [comment]
ok yes no problem I will look at it tomorrow. Thanks for all the info, that's very useful.

**Comment by soumagne** [comment]
OK I think that the tcp/rxm has the same issue but I cannot fix it at the OFI plugin layer. This will need to be fixed in the rxm or tcp provider most likely. Since everything is connection less from the OFI point of view, we don't get notified when a client drops, therefore the entry in the AV still remains. Having the client side explicitly removing the entry does not seem to signal the server either. So that's definitely a bug in the rxm provider that should be reported.

**Comment by carns** [comment]
@soumagne can you comment on this on the libfabric repo?

https://github.com/ofiwg/libfabric/issues/5381#issuecomment-546492453

**Comment by soumagne** [comment]
Fixed in f3bff206e1c5edbb6174a6a9bf9af614fefd51c9
